docs/csharp/language-reference/operators/boolean-logical-operators.md

```diff
@@ -0,0 +1,159 @@
+---
+title: "Boolean logical operators - C# Reference"
+description: "Learn about C# operators that perform logical negation, conjunction (AND), and inclusive and exclusive disjunction (OR) operations with Boolean operands."
+ms.date: 04/05/2019
+author: pkulikov
+f1_keywords: 
+  - "!_CSharpKeyword"
+  - "&_CSharpKeyword"
+  - "^_CSharpKeyword"
+  - "|_CSharpKeyword"
+  - "&&_CSharpKeyword"
+  - "||_CSharpKeyword"
+helpviewer_keywords: 
+  - "logical operators [C#]"
+  - "short-circuiting operators [C#]"
+  - "logical negation operator [C#]"
+  - "NOT operator [C#]"
+  - "! operator [C#]"
+  - "AND operator [C#]"
+  - "ampersand operator [C#]"
+  - "conjunction operator [C#]"
+  - "& operator [C#]"
+  - "exclusive OR operator [C#]"
+  - "XOR operator [C#]"
+  - "^ operator [C#]"
+  - "OR operator [C#]"
+  - "disjunction operator [C#]"
+  - "| operator [C#]"
+  - "conditional logical AND operator [C#]"
+  - "short-circuiting logical AND operator [C#]"
+  - "&& operator [C#]"
+  - "conditional logical OR operator [C#]"
+  - "short-circuiting logical OR operator [C#]"
+  - "|| operator [C#]"
+---
+# Boolean logical operators (C# Reference)
+
+The following operators perform logical operations with [bool](../keywords/bool.md) operands:
+
+- Unary [`!` (logical negation)](#logical-negation-operator-) operator.
+- Binary [`&` (logical AND)](#logical-and-operator-), [`|` (logical OR)](#logical-or-operator-), and [`^` (logical exclusive OR)](#logical-exclusive-or-operator-) operators. Those operators always evaluate both operands.
+- Binary [`&&` (conditional logical AND)](#conditional-logical-and-operator-) and [`||` (conditional logical OR)](#conditional-logical-or-operator-) operators. Those operators evaluate the second operand only if it's necessary.
+
+The `&` and `|` operators are also defined for `bool?` operands. For more information, see [The bool? type](../../programming-guide/nullable-types/using-nullable-types.md#the-bool-type) section of the [Using nullable types](../../programming-guide/nullable-types/using-nullable-types.md) article.
+
+For the operands of [integral](../keywords/integral-types-table.md) types, the `&`, `|`, and `^` operators perform bitwise logical operations.
+
+## Logical negation operator !
+
+The unary `!` operator computes logical negation of its operand. That is, it produces `true`, if the operand evaluates to `false`, and `false`, if the operand evaluates to `true`:
+
+[!code-csharp-interactive[logical negation](~/samples/snippets/csharp/language-reference/operators/LogicalOperators.cs#Negation)]
+
+## Logical AND operator &amp;
+
+The `&` operator computes the logical AND of its operands. The result of `x & y` is `true` if both `x` and `y` evaluate to `true`. Otherwise, the result is `false`.
+
+The `&` operator evaluates both operands even if the first operand evaluates to `false`, so that the result must be `false` regardless of the value of the second operand. The following example demonstrates that behavior:
```

Comment:
This is a super-nit, but it's more accurate to say it as follows:

> ...even if the first operand doesn't evaluate to `true`, so that …

That difference only matters for types that override operator `true` and operator `false` *and* could be in a state where both operators return `false`.

The pertinent clause in the spec is [here](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#conditional-logical-operators). See the two bullets following the sample.

Reply:
@BillWagner I like this nit-picking; also your comment confused me a bit.

According to the spec `x && y` is evaluated as `T.false(x) ? x : T.&(x, y)`. 

First, the `true` operator doesn't participate at all in definition and we cannot refer to it then. I assume when you say
>`x` doesn't evaluate to `true`

you talk about the `true` operator and mean that "`x` is not definitely true", or `T.true(x)` returns `false`. Do I understand you correctly?

Second, back to `x && y` as `T.false(x) ? x : T.&(x, y)`. I read that definition as 
> If the first operand is definitely false, do short-circuit. Otherwise, call the `&` operator (which means the second operator is evaluated as well).

So, when we say that the second operand is evaluated always, we emphasize that "always" means even if the first operand is definitely false (so even if there is condition for short-circuiting). I think that "the first operand is definitely false" can be rephrased as "the first operand evaluates to `false`".

If there is a flaw in my reasoning, please let me know. I remember I've looked at that spec and tried to build a sentence with keeping user-defined conditional operators in mind, but might have missed or misinterpreted something.

Reply:
You're analysis is correct. I'll resolve this conversation.

I had started editing and suggesting, then pulled out my copy of the spec as I got farther along. I was trying to make small edits without checking. That's always a mistake.